batch-l2.md

```diff
@@ -0,0 +1,1494 @@
+This example configures four retry strategies:
+
+```ts
+const jobDefn = new batch.EcsJobDefinition(this, 'JobDefn', {
+   containerDefinition: new batch.ContainerDefinition(this, 'containerDefn', {
+    image: ecs.ContainerImage.fromRegistry('public.ecr.aws/amazonlinux/amazonlinux:latest'),
+    memoryLimitMiB: 2048,
+    compatibility: Compatibility.EC2, // specify FARGATE for Fargate workflows
+  }),
+  attempts: 5,
+  retryStrategies: [{
+    action: Action.EXIT,
+    onExitCode: '40*', // matches 40, 400, 404, 40230498, etc
+  }],
+});
+
+jobDefn.addRetryStrategy({
+  action: Action.RETRY,
+  onStatusReason: StatusReason.SPOT_INSTANCE_RECLAIMED,
+});
+
+jobDefn.addRetryStrategy({
+   action: Action.EXIT,
+   onReason: Reason.CANNOT_PULL_CONTAINER,
+});
+
+jobDefn.addRetryStrategy({
+  action: Action.RETRY,
+  onStatusReason: StatusReason.custom('Some other reason*'),
+});
+```
+
+### Running single-container ECS workflows
+
+Batch supports ECS workflows natively. This examples creates a `JobDefinition` that runs a single container with ECS:
+
+```ts
+const jobDefn = new batch.EcsJobDefinition(this, 'JobDefn', {
+  containerDefinition: new batch.ContainerDefinition(this, 'containerDefn', {
+    image: ecs.ContainerImage.fromRegistry('public.ecr.aws/amazonlinux/amazonlinux:latest'),
+    memoryLimitMiB: 2048,
+    compatibility: Compatibility.EC2, // specify FARGATE for Fargate workflows
+  }),
+});
+```
+
+For workflows that need persistent storage, batch supports mounting `Volume`s to the container.
+You can both provision the volume and mount it to the container in a single operation:
+
+```ts
+jobDefn.addVolume({
+  name: 'myVolume',
+  efsVolumeConfiguration: {
+    fileSystem: myFileSystem
+  },
+  containerPath: '/Volumes/myVolume',
+});
+```
+
+### Running Kubernetes Workflows:
+
+Batch also supports running workflows on EKS. The following example creates a `JobDefinition` that runs on EKS:
+
+```ts
+const eksContainer = new batch.EksContainerDefinition(this, 'myEksContainer', {
+   image: ContainerImage.fromRegistry('my-registry/my-image:latest'),
+});
+
+const eksPod = new batch.EksPod(this, 'myEksPod', {
+   containers: [eksContainer],
+});
+
+new batch.EksJobDefinition(this, 'myEksJobDefn', {
+   pod: eksPod,
+});
+
+// alternative using convienience methods:
+const pod = new batch.EksPod(this, 'myEksPod');
+const jobDefn = new batch.EksJobDefinition(this, 'myEksJobDefn', {
+  pod,
+});
+
+jobDefn.addContainer('myEksContainer', { // adds to the Pod
+  image: ecs.ContainerImage.fromRegistry('my-registry/my-image:latest'),
+});
+```
+
+You can mount `Volume`s to these containers in a single operation:
+
+```ts
+jobDefn.addEmptyDirVolume({
+  name: 'emptyDir',
+  mountPath: '/Volumes/emptyDir',
+});
+jobDefn.addHostPathVolume({
+  name: 'hostPath',
+  hostPath: '/sys', 
+  mountPath: '/Volumes/hostPath',
+});
+jobDefn.addSecretVolume({
+  name: 'secret', 
+  secret: new Secret(this, 'mySecret'),
+  mountPath: '/Volumes/secret',
```

Comment:
https://kubernetes.io/docs/concepts/storage/volumes/#secret
https://docs.aws.amazon.com/batch/latest/APIReference/API_EksSecret.html

Batch supports secret volumes in kube pods.  That means the actual secret value is stored in Kube, life-cycled and authorized by the customer within Kube. The Batch job definition just points to that secret location; setting or specifying secret values is not support in Batch.

Reply:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-batch-jobdefinition-eksvolume.html
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-batch-jobdefinition-secret.html

I used the Secret class here because the EKS Volume Secret requires you to specify a Secrets Manager or SSM Parameter ARN as well as the secret name. The CDK Secret construct exposes both it's ARN and name, so it seems natural to use it here. 

How is the Secret ARN used by the volume? I thought that the Volume would pull the relevant information from the Secret to set it up, since you have to specify the ARN.

Reply:
Discussed offline, the CFN types are wrong here